libs/back/partial-backup/src/main.ts

```diff
@@ -0,0 +1,310 @@
+  const recursiveExtract = async (root: structItem) => {
+    print(`TRAVERSING ${root.path}`);
+    if (!traversals[root.type]) {
+      // console.log(`TRAVERSAL NOT AVAILABLE FOR ${root.type}`);
+      return;
+    }
+    for (const item of traversals[root.type]) {
+      const getter = pipelines[item.type]?.getter;
+      if (!getter) {
+        // console.log(`MISSING GETTER FOR ${item.type}`);
+        continue;
+      }
+      const objects = await getter?.(item.foreignKey, root.obj[item.localKey]);
+      const filteredObjects = objects?.filter(obj => {
+        if (alreadyFetched[obj.id]) {
+          return false;
+        }
+        alreadyFetched[obj.id] = {
+          type: item.type,
+          obj
+        };
+        return true;
+      });
+      if (filteredObjects?.length) {
+        const subRoots: structItem[] = filteredObjects.map(obj => ({
+          type: item.type,
+          obj,
+          path: `${root.path}\n${">".repeat(root.depth + 1)}${item.type}(${
+            obj.id
+          })`,
+          depth: root.depth + 1,
+          children: []
+        }));
+        root.children = [...root.children, ...subRoots];
+      }
+    }
+    for (const subRoot of root.children) {
+      await recursiveExtract(subRoot);
+    }
+  };
+  await recursiveExtract(bsdRoot);
+
+  /*
+    build a little recap object to know how many objects of which type have been loaded
+  */
+  const statsByType = {};
+  for (const id of Object.keys(alreadyFetched)) {
+    if (!statsByType[alreadyFetched[id].type]) {
+      statsByType[alreadyFetched[id].type] = 1;
+    } else {
+      statsByType[alreadyFetched[id].type] += 1;
+    }
+  }
+
+  print(`DUMP COMPLETE!`);
+  console.log("\n\nWhat will be copied :");
+  console.log(statsByType);
+
+  // console.log(statsByType);
+  const continueRes = await questionPromise(
+    "Do you want to write this to the destination database? (make sure it is empty and the schema is built) Y/N : "
+  );
+  if (
+    continueRes !== "y" &&
+    continueRes !== "Y" &&
+    continueRes.toLowerCase() !== "yes"
+  ) {
+    console.log("ABORTING");
+    return;
+  }
+
+  /*
+    Now we save the loaded objects to the destination DB.
+    Since there are some foreign key constraints, some objects have to be written before others
+    or the writing fails.
+    Since it's complicated to know the right order, I chose a more bruteforce approach:
+    - Try to write all the objects
+    - Remember which ones got saved
+    - Do it again with the ones that didn't get saved
+    - Stop when everything is saved
+
+    This has a risk of never ending, for example if there is a write error that is not
+    caused by a foreign key constraint (conflicting id of the db was not empty,
+    wrong format if the schema is different between source and destination, ...).
+    To avoid an infinite loop, I had a check that at least one object is saved on each iteration.
+    This way if we're stuck and nothing gets saved, we abort.
```

Comment:
Tu as essayé de tout créer dans une transaction en décalant le check des contraintes à la fin de la transaction ?
Avec quelque chose comme `SET CONSTRAINTS ALL DEFERRED;` à jouer en début de transaction.

Reply:
Alors c'est toute une histoire avec Prisma ça. J'avais déjà rencontré ce problème dans un précédent projet, et il n'a pas été réglé. Pour pouvoir utiliser `SET CONSTRAINTS ALL DEFERRED;` il faut d'abord que la contrainte soit "DEFERRABLE". Or prisma crée les contraintes en "NOT DEFERRABLE", il n'y a aucun moyen de le changer, et si tu le fais à la main dans la db ça crée des incohérences avec les migrations et ça pète tout. (cf. https://github.com/prisma/prisma/issues/8807). Donc en gros pas possible et ça rend un peu fou.